### PR TITLE
Fix slow performance for confusion matrix based metrics

### LINKED ISSUE #1267: Performance difference between `v0.9.3` <-> `v0.10.0`
## 🐛 Bug

Torchmetrics works on different speed accross v0.9.3 and v0.10.0

### To Reproduce

Same metric calculation on different versions installed environments.

> Results for `v0.9.3`

```
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                                                                                                                                                                                                             	|  Mean duration (s)	|  Num calls      	|  Total time (s) 	|  Percentage %   	|
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                                                                                                                                                                                                              	|  -              	|  1042           	|  17.113         	|  100 %          	|
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  [LightningDataModule]SegmentationDataModule.prepare_data                                                                                                                                                           	|  5.1728         	|  1              	|  5.1728         	|  30.228         	|
|  run_training_epoch                                                                                                                                                                                                 	|  4.6598         	|  1              	|  4.6598         	|  27.23          	|
|  [Strategy]SingleDeviceStrategy.validation_step                                                                                                                                                                     	|  0.28901        	|  12             	|  3.4681         	|  20.266         	|
|  [LightningDataModule]SegmentationDataModule.setup                                                                                                                                                                  	|  2.2177         	|  1              	|  2.2177         	|  12.959         	|
|  run_training_batch                                                                                                                                                                                                 	|  0.16763        	|  10             	|  1.6763         	|  9.7954         	|
|  [LightningModule]Model.optimizer_step                                                                                                                                                                             	|  0.16661        	|  10             	|  1.6661         	|  9.7361         	|
|  [TrainingEpochLoop].train_dataloader_next                                                                                                                                                                          	|  0.13849        	|  10             	|  1.3849         	|  8.0928         	|
|  [Strategy]SingleDeviceStrategy.backward                                                                                                                                                                            	|  0.077044       	|  10             	|  0.77044        	|  4.5021         	|
|  [Strategy]SingleDeviceStrategy.training_step                                                                                                                                                                       	|  0.056393       	|  10             	|  0.56393        	|  3.2954         	|
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

> Results for `v0.10.0`

```
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                                                                                                                                                                                                             	|  Mean duration (s)	|  Num calls      	|  Total time (s) 	|  Percentage %   	|
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                                                                                                                                                                                                              	|  -              	|  1042           	|  85.772         	|  100 %          	|
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  run_training_epoch                                                                                                                                                                                                 	|  62.396         	|  1              	|  62.396         	|  72.747         	|
|  [Strategy]SingleDeviceStrategy.validation_step                                                                                                                                                                     	|  3.159          	|  12             	|  37.908         	|  44.196         	|
|  run_training_batch                                                                                                                                                                                                 	|  3.6079         	|  10             	|  36.079         	|  42.064         	|
|  [LightningModule]Model.optimizer_step                                                                                                                                                                             	|  3.6072         	|  10             	|  36.072         	|  42.056         	|
|  [Strategy]SingleDeviceStrategy.training_step                                                                                                                                                                       	|  3.5262         	|  10             	|  35.262         	|  41.112         	|
|  [LightningDataModule]SegmentationDataModule.prepare_data                                                                                                                                                           	|  5.1832         	|  1              	|  5.1832         	|  6.043          	|
|  [LightningDataModule]SegmentationDataModule.setup                                                                                                                                                                  	|  2.2608         	|  1              	|  2.2608         	|  2.6358         	|
|  [TrainingEpochLoop].train_dataloader_next                                                                                                                                                                          	|  0.1303         	|  10             	|  1.303          	|  1.5191         	|
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

#### Code sample

Since its private project, I can only say the used metric names and configurations

For training:

```python

metric_params = {
    "num_classes": self.num_classes,
    "average": None,
    "mdmc_average": "samplewise",
}

train = tm.MetricCollection(
    {
        "IoU": tm.JaccardIndex(**metric_params),
        "DSC": tm.Dice(**metric_params),
    },
    prefix="Train/Seg/",
)

validation  tm.MetricCollection(
    {
        "IoU": tm.JaccardIndex(**metric_params),
        "DSC": tm.Dice(**metric_params),
        "Spec": tm.Specificity(**metric_params),
        "Sens": tm.Recall(**metric_params),
    },
    prefix="Val/Seg/",
)
```
### Expected behavior

More performant results.

### Environment

- TorchMetrics version `v0.9.3` and `v0.10.0`
- Python `3.8.10`, PyTorch `1.12.1+cu116`
- `Linux 5.15.0-1018-gcp #24~20.04.1-Ubuntu SMP Mon Sep 12 06:14:01 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

### LINKED ISSUE #1277: MatthewsCorrCoef has massive slow down on both CPU and especially GPU
## 🐛 Bug

It's extremely slow to compute the Matthew Correlation Coefficient since torchmetrics == 0.10.0.

### To Reproduce

During my testing, I am seeing massive slow down when computing the Matthew Correlation Coefficient, especially on the GPU (using pytorch-lightning to build, train and test a deep learning model's performance on this metric).

I have compiled a code sample (next section) and below shows the result over different versions of torchmetrics. 

ON CPU:
torchmetrics 0.7.3: 0.93961 s
torchmetrics 0.8.0: 0.93549 s
torchmetrics 0.8.2: 0.94494 s
torchmetrics 0.9.0: 0.92856 s
torchmetrics 0.9.2: 0.93682 s
torchmetrics 0.10.0: 1.10903 s

ON GPU:
torchmetrics 0.7.3: 0.11444 s
torchmetrics 0.8.0: 0.11682 s
torchmetrics 0.8.2: 0.11425 s
torchmetrics 0.9.0: 0.11433 s
torchmetrics 0.9.2: 0.11410 s
torchmetrics 0.10.0: 359.30208 s

So yeah testing over thousands of batches now takes almost a week for me to complete :rofl: Please take a look at this soon.

#### Code sample

```
from tqdm.auto import tqdm
import time
import torch
from torchmetrics import MatthewsCorrCoef

torch.manual_seed(1)

b, h, w = 10, 1080, 1920
device = "cpu"

def generate(b, h, w):
    prob = torch.rand(b, h, w).to(device)
    truth = torch.randint(0, 2, (b, h, w)).to(device)
    return prob, truth

batches = []
for _ in range(10):
    batches.append(generate(b, h, w))

mcc = MatthewsCorrCoef(num_classes=2).to(device)

t1 = time.time()
for detections, targets in tqdm(batches):
    mcc.update(detections, targets)
print(f"{time.time() - t1:.5f}")
```

### Expected behavior

It's supposed to be faster in the range of ~0.1 second for GPU and ~0.9 second for CPU in this naive benchmark.

### Environment

- GPU: NVIDIA RTX 3090
- TorchMetrics version (`pip`): 0.10.0
- Python & PyTorch Version (e.g., 1.0): python=3.9, pytorch=1.12.1
- Any other relevant information such as OS (e.g., Linux): ubuntu18.04, (same behavior with 20.04)

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1267
Fixes #1277
After the refactor the underlying `_bincount` function that does a lot of the computations for confusion matrix based metrics was changed. I did some testing at the time and everything seemed to be fine. However, for large inputs the
implementation is really slow. 

Here is a direct comparison in colab:
https://colab.research.google.com/drive/18tGZj_dPria6NSwVOIgwPXO8mJBr21kc?usp=sharing

The results:
image
The TLDR:
* Regardless of batch size the new implementation is approx 2x slower on cpu
* For batch size 100 the new implementation is approx ~4x FASTER on gpu
* For batch size 10.000 the new implementation is approx ~30x SLOWER on gpu

The simple fix currently in this PR is to change it back to the old implementation. The alternative is that we have something like this in the code:
```python
def _bincount(x, minlength=minlength):
    ...
    if len(x) > N or not x.is_cuda:
        return torch.bincount(x, minlength=minlenght)
    else:
        z = torch.zeros(minlength, device=x.device, dtype=x.dtype)
        return z.index_add_(0, x, torch.ones_like(x))
```
where we have to set N based on some experimentation.
@Borda, @justusschock opinions?

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]
@SkafteNicki For now, I think sticking to the default here makes sense. I would be open for changing this conditionally in the future, but this would require more experimenting :)

**Comment by Borda** [comment]
> I would be open for changing this conditionally in the future, but this would require more experimenting :)

do you want to open an issue for it so we won't forget? :otter:

**Review by stancld — APPROVED** [review]
